{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The problem: You've just completed an amplicon sequencing run on the 454 instrument, but because the sequences are longer than any you've previously generated on 454, you suspect that you may have sequenced through your reverse primers and into non-biological sequence (e.g., sequencing adapters).\n",
    "\n",
    "The solution: You want to find your reverse PCR primer in each of the sequences, and remove that and all bases following it. To do this, you can use scikit-bio's [global nucleotide aligner](http://scikit-bio.org/docs/latest/generated/skbio.core.alignment.pairwise.global_pairwise_align_nucleotide.html#skbio.core.alignment.pairwise.global_pairwise_align_nucleotide). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import random\n",
    "\n",
    "from skbio.alignment import global_pairwise_align_nucleotide\n",
    "from skbio import DNA, SequenceCollection"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, we'll model some sequences. This is quick-and-dirty. Each sequence will contain some biological sequence (which is what we actually care about) with a mean/std length of 400/40, followed by one of four slightly different reverse primers (so representing a primer with 4-fold degeneracy), followed by some non-biological sequence with a mean/std length of 25/2. This is a reasonable representation of what we'd get off of the sequencing instrument: our reverse primer is somewhere in the sequence, but we don't know the exact start or end positions. *(Note that I'm not modeling any sort of sequencing error here, and the biological sequence is random, which is not representative of what we'd have in an amplicon sequencing run.)*\n",
    "\n",
    "Follow the inline comments for descriptions of each step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "sequences = []\n",
    "num_sequences = 50\n",
    "mean_biological_sequence_length = 400\n",
    "std_biological_sequence_length = 40\n",
    "mean_nonbiological_sequence_length = 25\n",
    "std_nonbiological_sequence_length = 2\n",
    "# imagine that we have four slightly different reverse primers\n",
    "reverse_primers = [DNA(\"ACCGTCGACCGTTAGGATT\"),\n",
    "                   DNA(\"ACCGTGGACCGTGAGGATT\"),\n",
    "                   DNA(\"ACCGTCGACCGTTAGGATT\"),\n",
    "                   DNA(\"ACCGTGGACCGTGAGGATT\")]\n",
    "\n",
    "for i in range(num_sequences):\n",
    "    # determine the length for the current biological sequence. if it's less than 1, make the length 0\n",
    "    biological_sequence_length = int(np.random.normal(mean_biological_sequence_length,\n",
    "                                                      std_biological_sequence_length))\n",
    "    if biological_sequence_length < 1:\n",
    "        biological_sequence_length = 0\n",
    "    # generate a random sequence of that length\n",
    "    biological_sequence = ''.join(np.random.choice(list('ACGT'),biological_sequence_length))\n",
    "    \n",
    "    \n",
    "    # determine the length for the current non-biological sequence. if it's less than 1, make the length 0\n",
    "    non_biological_sequence_length = int(np.random.normal(mean_nonbiological_sequence_length,\n",
    "                                                          std_nonbiological_sequence_length))\n",
    "    if non_biological_sequence_length < 1:\n",
    "        non_biological_sequence_length = 0\n",
    "    # generate a random sequence of that length\n",
    "    non_biological_sequence = ''.join(np.random.choice(list('ACGT'), non_biological_sequence_length))\n",
    "\n",
    "    # choose one of the four reverse primers at random\n",
    "    reverse_primer = random.choice(reverse_primers)\n",
    "    \n",
    "    # construct the observed sequence as the biological sequence, followed by the primer, followed by the \n",
    "    # non-biological sequence\n",
    "    observed_sequence = ''.join(map(str, [biological_sequence, reverse_primer, non_biological_sequence]))\n",
    "    seq_id = \"seq%d\" % i\n",
    "    \n",
    "    # append the result to the sequences list\n",
    "    sequences.append(DNA(observed_sequence, metadata={'id': seq_id}))\n",
    "\n",
    "# construct a skbio.SequenceCollection containing all of the sequences just generated\n",
    "sequences = SequenceCollection(sequences)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If we now want to view the sequences we just created, we can simply print the ``SequenceCollection`` object, which gives us a fasta-formatted string representation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DNA\n",
      "---------------------------------------------------------------------\n",
      "Metadata:\n",
      "    'id': 'seq0'\n",
      "Stats:\n",
      "    length: 429\n",
      "    has gaps: False\n",
      "    has degenerates: False\n",
      "    has non-degenerates: True\n",
      "    GC-content: 48.48%\n",
      "---------------------------------------------------------------------\n",
      "0   AGTCGTACGA GGCACCCTGA CCATATACCG TTTGGGAGAC GTTAAACGAA TCCAGGTGAC\n",
      "60  GGCAGTGCGG AATGTAGTAG GTGCTCCGGG ACCTAACGAT GTCTCCGGAG GGTCTTATCC\n",
      "...\n",
      "360 TATGACGCTA TGAATGTAGG GACCGTGGAC CGTGAGGATT TTACAGAACT CACTACCTAT\n",
      "420 GACTTAGCA\n"
     ]
    }
   ],
   "source": [
    "print(repr(sequences[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now to get to the problem at hand. How do we find the primer sequence in random sequence. The answer is with global alignment. If we align the first reverse primer to the first sequence, we get an [``skbio.Alignment``](http://scikit-bio.org/docs/latest/generated/skbio.core.alignment.Alignment.html#skbio.core.alignment.Alignment) object back. If we print that object, we again get a fasta-formatted string. This lets us see how the sequences aligned to each other.\n",
    "\n",
    "Notice that in this step we get an ``EfficencyWarning``. That's because scikit-bio currently only has a python implementation of global alignment, which is slow because it's a computationally complex algorithm. In the future, we'll have a C-based implementation which will be much faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function global_pairwise_align_nucleotide in module skbio.alignment._pairwise:\n",
      "\n",
      "global_pairwise_align_nucleotide(seq1, seq2, gap_open_penalty=5, gap_extend_penalty=2, match_score=1, mismatch_score=-2, substitution_matrix=None, penalize_terminal_gaps=False)\n",
      "    Globally align pair of nuc. seqs or alignments with Needleman-Wunsch\n",
      "    \n",
      "    State: Experimental as of 0.4.0.\n",
      "    \n",
      "    Parameters\n",
      "    ----------\n",
      "    seq1 : str, Sequence, or Alignment\n",
      "        The first unaligned sequence(s).\n",
      "    seq2 : str, Sequence, or Alignment\n",
      "        The second unaligned sequence(s).\n",
      "    gap_open_penalty : int or float, optional\n",
      "        Penalty for opening a gap (this is substracted from previous best\n",
      "        alignment score, so is typically positive).\n",
      "    gap_extend_penalty : int or float, optional\n",
      "        Penalty for extending a gap (this is substracted from previous best\n",
      "        alignment score, so is typically positive).\n",
      "    match_score : int or float, optional\n",
      "        The score to add for a match between a pair of bases (this is added\n",
      "        to the previous best alignment score, so is typically positive).\n",
      "    mismatch_score : int or float, optional\n",
      "        The score to add for a mismatch between a pair of bases (this is\n",
      "        added to the previous best alignment score, so is typically\n",
      "        negative).\n",
      "    substitution_matrix: 2D dict (or similar)\n",
      "        Lookup for substitution scores (these values are added to the\n",
      "        previous best alignment score). If provided, this overrides\n",
      "        ``match_score`` and ``mismatch_score``.\n",
      "    penalize_terminal_gaps: bool, optional\n",
      "        If True, will continue to penalize gaps even after one sequence has\n",
      "        been aligned through its end. This behavior is true Needleman-Wunsch\n",
      "        alignment, but results in (biologically irrelevant) artifacts when\n",
      "        the sequences being aligned are of different length. This is ``False``\n",
      "        by default, which is very likely to be the behavior you want in all or\n",
      "        nearly all cases.\n",
      "    \n",
      "    Returns\n",
      "    -------\n",
      "    skbio.Alignment\n",
      "        ``Alignment`` object containing the aligned sequences as well as\n",
      "        details about the alignment.\n",
      "    \n",
      "    See Also\n",
      "    --------\n",
      "    local_pairwise_align\n",
      "    local_pairwise_align_protein\n",
      "    local_pairwise_align_nucleotide\n",
      "    skbio.alignment.local_pairwise_align_ssw\n",
      "    global_pairwise_align\n",
      "    global_pairwise_align_protein\n",
      "    \n",
      "    Notes\n",
      "    -----\n",
      "    Default ``match_score``, ``mismatch_score``, ``gap_open_penalty`` and\n",
      "    ``gap_extend_penalty`` parameters are derived from the NCBI BLAST\n",
      "    Server [1]_.\n",
      "    \n",
      "    This function can be use to align either a pair of sequences, a pair of\n",
      "    alignments, or a sequence and an alignment.\n",
      "    \n",
      "    References\n",
      "    ----------\n",
      "    .. [1] http://blast.ncbi.nlm.nih.gov/Blast.cgi\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(global_pairwise_align_nucleotide)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'skbio.sequence._dna.DNA'>\n",
      "<class 'skbio.sequence._dna.DNA'>\n",
      ">0\n",
      "---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ACCGTCGACCGTTAGGATT-----------------------------\n",
      ">seq0\n",
      "AGTCGTACGAGGCACCCTGACCATATACCGTTTGGGAGACGTTAAACGAATCCAGGTGACGGCAGTGCGGAATGTAGTAGGTGCTCCGGGACCTAACGATGTCTCCGGAGGGTCTTATCCACTTGACAGTCTAGGCGATTTTTGCATATGCTGAATCTACTCCCCTGCACGTATCCCGACTGCCGAGAGCAAACGCGTAATGTATCCTCAATTCACGGTATGCATGCACTAGGATTACTATAGCAGATCCGCTCGGGAGTGTACTGAAATTAAATCGAATCTAGGGGCTATCTAGATTTGCGCAATGTAGGTAATGTCAGGTGAAACGGCATCGGCAAACATCGATCGCTTAAGAGCAACTATGACGCTATGAATGTAGGGACCGTGGACCGTGAGGATTTTACAGAACTCACTACCTATGACTTAGCA\n",
      "\n",
      "<class 'skbio.sequence._dna.DNA'>\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/caporaso/Dropbox/code/scikit-bio/skbio/alignment/_pairwise.py:549: EfficiencyWarning: You're using skbio's python implementation of Needleman-Wunsch alignment. This is known to be very slow (e.g., thousands of times slower than a native C implementation). We'll be adding a faster version soon (see https://github.com/biocore/scikit-bio/issues/254 to track progress on this).\n",
      "  \"to track progress on this).\", EfficiencyWarning)\n"
     ]
    }
   ],
   "source": [
    "print(type(reverse_primers[0]))\n",
    "print(type(sequences[0]))\n",
    "aln = global_pairwise_align_nucleotide(reverse_primers[0], sequences[0])\n",
    "print(aln)\n",
    "print(type(aln[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We next want to find the start position of the primer sequence in the sequencing product, which we can do using the ``gaps`` boolean vector of the first sequence in the alignment (to learn about ``gap_vector``). The following tells us where the first non-gap character in the primer alignment is, which is the position in the sequencing product where the primer match begins."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "381\n"
     ]
    }
   ],
   "source": [
    "gap_vector = aln[0].gaps()\n",
    "primer_start_index = (~gap_vector).nonzero()[0][0]\n",
    "print(primer_start_index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we can slice the original sequence through that position, and the result will be our sequencing product minus the reverse primer and the non-biological sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGTCGTACGAGGCACCCTGACCATATACCGTTTGGGAGACGTTAAACGAATCCAGGTGACGGCAGTGCGGAATGTAGTAGGTGCTCCGGGACCTAACGATGTCTCCGGAGGGTCTTATCCACTTGACAGTCTAGGCGATTTTTGCATATGCTGAATCTACTCCCCTGCACGTATCCCGACTGCCGAGAGCAAACGCGTAATGTATCCTCAATTCACGGTATGCATGCACTAGGATTACTATAGCAGATCCGCTCGGGAGTGTACTGAAATTAAATCGAATCTAGGGGCTATCTAGATTTGCGCAATGTAGGTAATGTCAGGTGAAACGGCATCGGCAAACATCGATCGCTTAAGAGCAACTATGACGCTATGAATGTAGGG\n"
     ]
    }
   ],
   "source": [
    "print(sequences[0][:primer_start_index])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, if we want to do this for all of the sequences, we can embed the above steps in a loop over the ``SequenceCollection``. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "trimmed_sequences = []\n",
    "\n",
    "for sequence in sequences:\n",
    "    aln = global_pairwise_align_nucleotide(reverse_primers[0], sequence)\n",
    "    gap_vector = aln[0].gaps()\n",
    "    primer_start_index = (~gap_vector).nonzero()[0][0]\n",
    "    trimmed_sequences.append(DNA(sequence[:primer_start_index], metadata={'id': sequence.metadata['id']}))\n",
    "\n",
    "trimmed_sequences = SequenceCollection(trimmed_sequences)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can then print the result, and we'll have acheived our goal."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">seq0\n",
      "AGTCGTACGAGGCACCCTGACCATATACCGTTTGGGAGACGTTAAACGAATCCAGGTGACGGCAGTGCGGAATGTAGTAGGTGCTCCGGGACCTAACGATGTCTCCGGAGGGTCTTATCCACTTGACAGTCTAGGCGATTTTTGCATATGCTGAATCTACTCCCCTGCACGTATCCCGACTGCCGAGAGCAAACGCGTAATGTATCCTCAATTCACGGTATGCATGCACTAGGATTACTATAGCAGATCCGCTCGGGAGTGTACTGAAATTAAATCGAATCTAGGGGCTATCTAGATTTGCGCAATGTAGGTAATGTCAGGTGAAACGGCATCGGCAAACATCGATCGCTTAAGAGCAACTATGACGCTATGAATGTAGGG\n",
      ">seq1\n",
      "CATCTACTGTCATTCCGATACTAATAGACAAACTCCTATGGGAACTCCTGCATACGAAGGGCATTCAGACTTACTGTTTCCCCTATGCGCTTCTGACAAACGCCCGTCTCACAACGTTGTGCTAAGGTGCCCGTTTTACTCCAACGCCACGGTAACGTGGCCTGAGGTCACATACCTTAGATAATCTGCCCGACCGTCAGTCTAGAGATGTCAGCTCAGCTTCATACCTAGAGTGTAAGAATGAACCGTTTTGGAAGAGCGTCGCAACACAAGAGGCTCAGGGTAGACTACGCGTGCTCTTATCCGAGGTTCGGTGAGACCACTCAAATTAATTGTAGTCGCTCGAGCCGACCACGCCACAGCTAAGGTCGCGATAGAACCAGG\n",
      ">seq2\n",
      "TTCAAGTGTATGACCGCGGCAATAGTATTCAAGTCCGCGCCTAGGTCTTGTGCTTAGTGCTAGTTGATAGCGGCTAACAGTACCTCGAAGAAGATAAAGCCGGACCACATCGGTCGTGCTAATGCGTTTATTACATTGTGCCCGATATGTGCATAGAATATCAGCGACAGCTGTGTGTGGTTCGAATCGCGTCGTATTCTCGACACTCCAGCTCTAAGCGAATCGCGTACCGGGGTGCTCGCATATGGCGAACCAGGCCCCAAGTCTCGGCTTCCCAGGGTATGCATCAATGGAGAAACGGCCACTGAGACAACTGATGCGCCGCGGGTTTGCGGCGCTATGA\n",
      ">seq3\n",
      "GTGACTTCCCCCATTCTGTTGGACTAGGGATCTTTAGCCCTGTTTATTCACACAATACGCGGGCAGTCACGTTTTCTCCCGCTGAAAGCCCCGAGTTCGAACGAGCCGCCGGGACAAGATGAGTGCTCATGCCACTAGATCGTGGAGGACCTAATCTCCGTACCCAGATGCAAAAGAAGGGTTCAATCGACTGCCGGCCAATGGACTCTGAAGTTGATACTCAACTAGTTCTGGTCGCTAACCCGCACGGCGTCACTAGGGCTCGGTAGCTCTGAGTTATAACCGGAACCATGTGCGACGCCCAGGCGGCCTAAGGCCTTTGCCAAATGTTTCCAGGTTAGGATTCCCTAGGACTCATGAATGAGGTTGAAACCCAAAATCCCGTTTAATGCATGCTGAGCTACCGTCT\n",
      ">seq4\n",
      "TCAGGGAGATAGGGTCCCGGCTCTCTAGCAAACCTTGATACAAAACAAATACATTCCCTTCTTAGGAGCGAGCAACACCTCGGGTCGAAGAAGGCACGCAACGATGCTCAGTGGTTAGTTGCGACCGTGCCGCCATAACGGTTCGTGTCCTCCCATTATATGCGAGTAGTATGACTTGTTTAAGCGGAATGCCTTTCAGGATATTTGTTGTGCTGTGGTTGCCACAACTAAACGCGGGGTCCTGCTAGTACTACGCTACTTGCCACACCTCGGGGTTATCGATGCGGAAGGGGCTTGTGAACTCTCCGGAATTGTGGGAACCACTATGCCGTACATGCGGGTGGTCGTTATAGTAT\n",
      ">seq5\n",
      "AACGGATCGGAGAGACTTCGCCGGCCCAACAAACGTACCTACAGAGCTGATAGATTGTGTCGCCGCTGGCCGTAGTTCCAAGGAGCAGGGGACCCCGTCGGTCGTACCTACGGGTGACGCGTTCCTAAGTTACGTACGTGAGTCCACGACCGGCAAAGAATAGATTCGGCACGACCCATTGTGGACGACTCCGTCTGCTGAAATTATCCAGAGCCCTATAGAACTCCCCCTTAAACACCAAACCAGCCGCGTCGCGGTTGAACCGCCGACAGGCCGTTCGAAAACTTACAAGGTGATCGCCTGATATAACACGGGTGTAAATAGTTTGGACACGGGTATCCACGCCCTTTCTGCGAAGGTATCGCATACGTTCCTTCCGG\n",
      ">seq6\n",
      "AGGCGCACGCTGCAACTCGTGCTGTCGCCACATATTTCGTTAAACGATTATTCATTGCTTTCTCGCTACCGTTTACACACTCACTGCGAGAGCTCTGGTAAGGAATAAAACTCAGACGGTAGAAGCTGGGGGGCTATTTTGACGTAGGTGCATGGAGAGCGAAATCGGAGTTAACTTTCCTTTGTGAGCAGCGCTACGATCAGTACGTGGACTGCATCGGGGTGTATACTCGGGCCCCGGACCACCGAACATCATGCCTTTACCGTCAAGTAATTACATGAAAACCCTTATCTCCTTAAAAGCTCGCCCTGCCCTAAGGGAAAAGGCACATGGGTTTTAACGTCCGTACCTCAAGACTTGGTCAAAGTCATTCTGATAAAGGTCCTG\n",
      ">seq7\n",
      "ATCTGGCCTGGGGTTATATTTACGGTCTATATAGCCCATTACGTGATGTGGGAAAAGGCACGTTACTGGCGATGAACTGTATAACACTATGTATAACTTGCATTGCGGCTCATTGCATTCCTAACAGTTTCATTAACATGAGAGCAATTACGGGTCTTATCTTTGTAACTTACTATGTAAGCCCCCGAGGAGCAGCGGGTCATTGCTCGATGAGGCGAAGCAAAAGGAGCTGCGGACGAGATGTGTGCAGATCGCGAGGGGGCCGTGACAGGAGTCGAGGTTGGTAACAGCGCCGAAGCCTTCTAGGCCATTGGCGTACTTTACCAGCGTCGGTTAAAGCCGACTCCCGCAGTAGGATTCGCGGCTCCTTCTTCCATGTCACATATAAAGTGGTACCAGTATGGATCCCCATA\n",
      ">seq8\n",
      "ATGCACGGGTACAGCATTGGTTACCTGGTACCTGCAGCTGTTTGACTACAACTACCACTCGCTGAGCCAGGACTACAGTGGGACATTATGTTGCAGCGCCTGAGACAGTAATACTATTCCTGCGTTATAAATAGGGAAGAGTTACGGTACCTATATTACTGACCGGAAGTTAGAGTATCTCCTCTTCCAATTTAGCTTCCTGCAGAGGTACTTGTGTGCTCGGCAGTACTAGGTGCGTTGATTAGTGCAGTATCGTCTGTCCACTTGCCTCTAGGTGGCCCGACAGGAGGTGTGCGGTTAAACGTTATCAAGGGATCATTGCTTTTCTGGCGAAACGTACATAACTGCCGCCCGTAGCATAGCTGTGTTTCATCCAGATAAGTCTTTAGATCCAACCCTCGACCAATTTCGGCACAGGACGGGGCCACTCGGCTTGGCTATACGGGAGTCACCCGCCATTTCA\n",
      ">seq9\n",
      "CCCTGTCACACTGTATGCTACATTAACAGCATGACGGAGAGCGCAATCAACCCCCTTTTGAACACCGTTGATACCAAGGTCCAACGTGCCCCAGAACGAACTGGCTTTAGTTCCAACAGTCGTCTTACTGCCCCGGTCTCCCTAGCATACTTCATGCTCCGGCAGCTTACAATTGGATAGCGGCTCGCATACTGGGCTGTAACCATAGCTGCTGATATCTGTCCTGGTGTGAGAAATAGCGGCGTGATGAGCTGTACTGGGCTGTGCAGGTGGCTGCGAAGGGTGTAACCGCCGCTGTGAAGAATAGAAGCAGAGCGAATACGGATAAGTACCGCGTGTTTACGATTGTGCATGACCGGGTTAGCATATGCTTATGGACATGCAGCACTGCACTCGTGTATCTGCGAGTTAGAGTTAGT\n",
      ">seq10\n",
      "TTCGCCCTTGTGCGATCCTAGAACAAGTTCCATATCTCTGATTGGCACGCTCTTATGAAAATGAACCACCGGCTTGGTCAATGGACTCCCAACGACTATCGATAGTAGGCTGACGTTATAGGATTAATGGGCTGTGGTGTGCCAAAGACATATAGCGCTATCCGCTCAACAGCCGGGCGGGCTTGGAACAGCGCAGACCTCTTCCTCATCCAGCTAGGCTAGTAAGCTTACGATCGACGAGCCGCTCTGGCAGACTGCGTCCCGTGGCCCTTTACACTGTGGACGGGGACTGGAGAGGTGGTAATTAGACTCGAATCGAACATATGCAGATAGCATTGTGTTCAATACTTCTTGGAGAAGAAACGCTCAATGTTCTGGCGCCGGCGCCTAAATTAGCAAGCCTCCGGTCGGATTCATC\n",
      ">seq11\n",
      "CAGACAAATCCACATAGTATAGTTTCGAGTAGTCCTTATAAGATACTGCAGGTAGGCTGCCGCGCTGTCATGGCCGATTCCATAACATGGGACAACTACGCAGCCGCTCACTGTGAATAGCACCTCGTATCACATACTATCTAGCTAGTAAGGTACAAAAGTTATCCATTGCGAGCCGGCGGAGCGCCACGCTGAAGTAAGGATCTAGATTACAACTAAGCGTACTTAGGACCTCCTGGGGTGCCACGCCCGAACCGAGGGCTCAGGAAAATGAATGCATCCTCGCCAGCGAGTATTAGACAACACTTCGAAGTAAAATTAGAATTGTCCGCACAATATATACGCCCACATTGTGCAATCAGCTCCAATCTGCAGAACTCTCAGCCAGGGAGGCGATT\n",
      ">seq12\n",
      "TAGGGATGTTATATGCATATGACTTCATTAACCCGCAGATTCTTATGACTCCCAGTTACCAGTTACTTCTCGTACACAGCGGGGTTATGGATTGGGCTCCGTGCGAACTAAAAGAGACACAGAGAAGAATAGTGCAGTAACTGCGACGACGGCGTCGGTTATGCGGCACTATGCTGCTGCTTCGTGCGAAGAAGGTCGGACAACCTTAATTATCTTGCCTGTTTCAACAGCCGGCTTAGGATAGGAAGCAGCATCGATCAGAGAGACGAACATGAAGCTGAAAGAACGAAGACATCTTGAGTAATATTACACATTAGCGCTTCTCCGGAATATCGTACGGGACGGCGCGCAATAAGTGGTCGGTTGTCCCACAGTCAATAGTCGCTCGCAAGCCATGCCGAGTCCTACTCTACCAGAAACTT\n",
      ">seq13\n",
      "CTTCCTTCCCCCATATCGTTCACGCCTGTGCCTTAGGATAGGATACAGTAGCGAAAGCCGCCAACTCCAAGGGGAGCATCGGTTAGTCGCAATATCATTTAACTGGAACAGGCTCAAGTCCCGTCTCTCCCACAGTAGACTTTTAATTCCAAAACAGTCGTACCTACTTTTCCACGTCTGATTTCTGCACTTGCAACTGTATAGCCAATCGTGAACGAATCTGGCCCTTTATAGCTGCGTCAAACAGCATTGTCTTGACCTGGTTTCTACTATTGGCAAACAAGCATTTCATATCGATCGTTCTCCATGTCGGTGAAGGTTTAAGC\n",
      ">seq14\n",
      "CAGATCCCATCATGATTCGTTCGCGTCTTCTATCCCTAAGTCCTCATCGACTCCATGCACTATCAAGTAGTCGGTGCTCGCGGATATCTGCCATCCTATAGCGACGCTCTCTTCGGACTTGTATCTGCGTGCGATTACTTCATAAGTACAAGCCGGCTGAATTGTACTTGATCGATGACCAGAGTTCGCGAACAAAGCAGCCCCGACCGGAATAGGGGTGGTCACGGTTATCGATTAAAAGTGCCCCTCAGAGTGCCCTCCCCCATAGTTTCGTACGACCATAACGTTTAAGCTTGACTGCGCTAGTGTTTTCACCGCGCG\n",
      ">seq15\n",
      "TTAGTCCAGGGGGTTTGCCTTAAGGGTTAGTCTAGCGACGCCTCATGAGCCAACACACACGTGGAAGTCGAGTGGCTAAATACGTCAGCCGACGGCTCTAAACGCGACTTTTATAACCTGAATGTACACGCCCCCAGGAACTGCCCGTCGTCGAAAGTATTTAGGTCATAGGCTTTGCAGTGCAGTTAGTGGGGCCCGAGTAATTCTCGCCCACCCACTGGGTCCCGGTGACCTCTGTAGTATGCGTGTGCAGGACCGCCATTGAGCGGATAACAGCAGGCAACTCTTGGGGGCCTTTAACCACTCATCACAACTACTTCGCTCGGAATGGATGCCTGATATTTGACTATTCATACGACTCTGTCGATTGCTACCCCTAAACGGCACATTCCCGGGATTTCCTGAATTATCACTGGACAGAATTAAGTTCGCC\n",
      ">seq16\n",
      "GCGCTCGCGTATCGGTTCTCACCACCCTGGAAAGGCTCTGCGATGATCAACCACGGGCATTCTGGGATGGCTATATTACTAGAGCACGCGCCAAGCCCCGCGGAACTATTTTGGTTGAATTTTCCGTCGCACGGATCAATCAATATCAGGTGAAAAAACGCGAAAGGGAGGTACCGATATAGTGGGTAACCCGCGTCTGTGCGTACTATTTGTCTCCATTCGGCACTATTCGACAAAAGGCTAAGTGGCTAACGTGGCTCTATCTACCTGCGATAGTCGATAGGACACATAACTGGCCACGATCGGATCATCCCTCTTTCCTACGTCTCACTGGGGCAACGGAGGGCTCCGTACCCA\n",
      ">seq17\n",
      "AACCTCTGGGCGGGGTGGCGCTCGAAGGTGGCAAGACAGATAGGCTCCAGGCGCTTGAGTAGCCACTCAGCCACCCTCCGACAATTCAGAAAACTTCGCGGATCCCAAACCCAACCGTGAAAAGGCCAATGGCAGATGATCGGTTCCTCTTTGGCGTCCTTATAAATTAAGGGAGGGAGCGTAGGAGAAGGACTCAACGTGCAGCTGTTCTTTTGTCGACAGTGATGAAATCAAGTATATCGGCATTGCGTGCAATAGCCCGGCTAGTTATAGTTAGCCGGCGATTCTACTACAGTGACACCATGCTACGGAGATGTGCTACTCGCACTACAATAAGATCACTCGGCGTGCGGAAGACTCCGGAGTTGGGTTTGAATCAAGTAGATTACAATCTTCTTCAAATCACGGTGTGAATTATGGGAC\n",
      ">seq18\n",
      "CGTGCCTTACTACCCTTGTGTCTTGTAAAAAGCGCTCGGCAGATTTTTCGACTTATCATAGGCCTTTCAATTTGCACACCCAAAGGGATAAGCCCGAGTTGGCCGTGAATAAATTCTTTATTGCGCGCGATCAAAAGCGGTATTTTAAAATGCAGATCGCCCCGAAGAAAAGTATAAGCGTTGTCCGTCGCCAGCTGCCCAGGAAGCGCTCTGGCGGGAGATGTAACATTCTTATGGCTCTGTATGTTCAATTATCTATTTGACATTTGTTTGGAATACCAAGGCAGCCGAAACCGCATGCCCCCAAAGATTACCTAATTGTGAAGAGAAATTGGCAGCGCGACGCCGTCGCTTCGATCGGCTAGCAACAGTAGTA\n",
      ">seq19\n",
      "TGCATGGGCCCAACAGTGGCGTATTTTCGAAGAAACAACCCCTTATCATCCCCTTCTCTAAAAGTCTTTCAACGACGTTCCGGGTTACGCCCTCCCAATGCAGGGCACTGTCGGCTAGTGACACATCCAACGACATCGGCATATAAAGACCCCAAACGTCGTTTGTGGAGGCCGGAACAGATCACAAGCTGCACATCTCCTGCGATTTTCCTTACGCCTGGTACGGAACCAGTTAGTCCTGTGGTTGGCACTGCGATTGGACAGGGATGACATGGGTCGCAAGTTTATGGGGTTTGAATGAATACGCCGCACCAGGGCCGAGACTGGCAGCAGACTACTGGAGACACGCCCGGCAACTCTCTCCGGCCAACCTCCTCCAGGGAAGGGTACCAGACTAGTTTGTCGCCATATTCGACGCACGTATC\n",
      ">seq20\n",
      "ACACGACAATAATCATTGCCCGAACATTTGAGTTCGGACATTTTTGAGCCCGCTGCACTGTGTGCACATTCGGATCGCAGGATGAACGATAGGCTACACTTGAAAAAATGCTATAATTCTTAAGAGACGCATTAGGTTTCCCGTTCCTAAGTTGCGGGCACATTTTCACCATGGTTGACTGGTAGTAACCCAAGCCGTTTCTTCACGTTGCACACAAAGGAGGCGCTGACTAAGGCGCCCGCTCTTGGGAAGCGCAAAGGCTGGAGTTCCCGCTCGTCCACGTACCACTCATATCGATGTCAAACGCTATGAATTTGCGATACTAATAACACACATACGAGGCCCGAGTTAACATGATAGTTCATTGAGCGACCTGTTCGCTGAACCGTCTGCGCCTTTTGTCACCT\n",
      ">seq21\n",
      "TCTCGGCGAATCCTGGGTTGTCGCACAAGACGTGTAGGTATATGGCTATTCTACTTTAGATCCTTGTGTGTCGGGCTTGGAACTTGTACCGGCACTTTTCGGGATCTCGGTCCGTGAAACTTCTCTGTAGGTCGATTAACTACTGTACTATTGCGTCGACAGTCACGGTTAATAGGTAATCCAAGTTAAGGCCATACGTCAACGATTGTCTTGCACTTTGTCGAAGGACCACGGCCCCAAGTAGTTCTGACTAACCCTCGTGGTTTCGACTTTGTAGACGCGACCCCCTGGTTTCTACCTTATCAGTGCCACCACTAAGATTGTTTGCATCCTGTCCCGAGATGGGCTAAAAATCGCCTCTAAGGGCCCCTTTTCGATGTTGGTCCCCGAGGCTCGGTTGCGCAGGGCCCCAACTTCGTCCAG\n",
      ">seq22\n",
      "CACCTATCCACCACGACCTGGTCTAAGCTGAGGCCATCGGTGACCAGGGGCTGGGCCCTAGAATCTAATCAACGAAGGAGATGTGTCCAGCAATCGACTGATTCGCAGTTGATAGTGCACAATGGAGTACGACCCTCGGAAGGGGTCATCCAATGATGCAGGAAGCCCCTCCTTCCTGCGTCGCGCGTAACAACGAATCTGACCATGATTCTAAGATCGCAGAGCAATCGTGTTCACTCTGCCTTAGCTTCAGGACAGATCAATCGGCTGTCAAAAGGACGTGGCGAGCTTACCCAGAGTGTCCCCACGTGTATACATTCTACGGACTATTAAAATCCACAAGTTATGGGGGTGGTGTCGTGCTTGAGGTACGTTAGTACACGTGGTGA\n",
      ">seq23\n",
      "CCGGAGCCGTAAGGTGAATTCAGGGTAAGTCGCTATAGAATACTAATGCAATTGCGCGCCACCACGTGTCTTTCTCCCCCACCCCGGTGCACAAGGAATTATATGCACAACCAACTGGTCCCAGTCACCACTAGCACGCTTCGAAGGAGACTGGGGCCGGAGGACAACTGACCCTGTCGGCTGGAATCGTTCAATACCAGGGACCGCGATCACCTGCCTTAGACGTAAAACTGCTTGGACTGCGGCGTGCGTCATTGACACCAAATGGAGCAATTCGCGCGTATGTGCTTTAACTGCCTTAAATACTTCATTTCATCCCTAATTCTAATGCCAAGCTTTACTTACCCCTCGCATTCGATTTCGGAAACAGAGCACGGTCGATCGTCTACGTATTTCCATGGGGGAGTTCCTAGCGTGCGGCGT\n",
      ">seq24\n",
      "GGAAACGCTTCGGCTCAGTCGGTGCACTAGTACGAATCTAGTGCATAATGGCTGCGACTGTGTGATACGTCGCCATCAATACTGATGCCTGACTTCCATCCATTCCCCCAAGACCTACCTCAATCGCCACGTCTGGCCCCCATCAGCTCGGGCCGTACCCCCTGTAAACCGCCCGTCGTAAAGAATAGGTCTGTTGAAAAGCCCAGAGATCGCTGGCTCCTACGCACAATCAGTTTGCTGAAAAGTTAAACAATAATTATGTACACGGCTAACCCTTAGTTTCGCGATCCAAGACTGGCCACATCTCCCATATCCGCGGAGTAGTAGGGCTTATTCCCATGCAATGAATGGCACATGGGCTAGAAGAACCCACGAACTTAATGCATTGATTCGT\n",
      ">seq25\n",
      "GAGATAGTTTATCTGGCCTCTGATCCATATGAGCTATGCGCCCTCCCTTCGAGCATTTGTCCGGGGGGTTCACGCTCGAAGGACAAAATTGCGCTTGACGGGATTGGCATAATTCCGTACTTTGTGATTTGTTGGTCCCGCGCCAGAGGTGGGTGTAAAGGAACGGCGTATGCCCTTGTCACCAATCTAACAACTCCGCTAGGGACACACCACGAGTACCCAGAATTTTGATAAGCTTTCTTCACATCCGAATCGTTTCACCCAGTACGACTAAGTTCCTAGGTTCTTCGTGACCGCGTACTGCTTGAGGCAAGCGGGGTTATGTGTAAGCCTAGAATTTATTCCAGATCGAA\n",
      ">seq26\n",
      "GGCCACCTAAGAATCCGTTTAACCGGACCGGACTAGCGAAGAGTTGACGGTGGGGCCCCCGGTATTCACCATAAGAACGAAGGACGCCATGCCATACTTCAGAGAGTTTGTCTGAGCTGGGTCCCTCAAGTTTGGATGGAGCTTGGTTACAACTTATTCTGTTCCCGTGCGCGTGTTTGTGCTGGCATGAGTGGCCGGAATGGTGTTGGCCGTACGGGAGTGGTTTTTGGAGGATGTAACAATAGCACTTAGTTGGTCTGGCCCTACGGAGCTTTGCTGATTTAACGGGAGATATGGTAAAGAATTATAGATTCTCGGGTAGCTATTCCCCCCGAGCGGAGACTCATAGGTTAGATAGATCGGGAGTGCTGCTAAGATAACCCCGGATGGCATGAATTTCCTAGATTAGGGTCAATAATAAATTGTTGTGTTTACGGACTCCTTGGCTG\n",
      ">seq27\n",
      "CGGTTAAGGGTCATAAAGCGGTCGCAGAATTACATGTTAACCGGTATACGTCGGACAGAACGTTTCGAATGTGGAGAAGAGTTTATGGAACTATCCGAGCCCAGACCCCGACAGCGATCCACCACAGATGTTTACCCTGACTCTGGCCACAAGCGCCCATAAGTACACGTTTCGACATTCAGTGTGGACGGCTTTGAACGCCGAAGAGAACTCTGGGACCAGTCTCTGGGCGAGACAGCTGGCTAAATTTCGCGTCACTAATTACGTCACGAAGATTACTGCTGGTTCTTTTCGGTTTTACCGACACTGTGATTCATACATTAAGGTGCGTACGAAATAGGGTTGGACTCAATAGTAGAAGCTAAGTGCGGTTCCACCCTCTCCAGTTTAGTACAGGATCGACTTCATGTCGCGGTCT\n",
      ">seq28\n",
      "CTCCTCTTTGCAGGGATGTGTCCGGACTGGCTTCTTAAGCCTAGACCGTTGTAGACACAGTAGATGAAACTCTAATGCACTGTCAGTGGCTCTTCCCATCGAGCAGCTTCTGTAAGAAAATGCACCGGTAGAGGGTATGAATGGTGTAAGAAAGTCGCCTTCCCCTCGCCGGCCCAGTCATGATACCTCTAAAACCATCCAGGGCGGGTCTGGGAAGGACGCTGGGTGTCATTACCAGCAATACAAGCCTCGGGTCTAGGTCGCCAATGCATTATATTAGAACGAATGAGAAGTAACAGATTCGTGGAGGGACTCAGCAGGGATTGAGTGGCACCAGATCATAGAAGGAGTTGTTCACGTTACAGCG\n",
      ">seq29\n",
      "AGATCGTACACACTGTATTCTTTTCCCTAGGTCAATAGATGAGTAGCGTGATTATACCGCCGAGGTGGATCGCTCTGCCACCAAGTCCCTAAGCTTTGAAGCAGAGCGTTCGACTACCCCGCTCAACAAATCGCCACTGCTCACCCTCGACTTCAGATAGTACCGAAAGCAGCAAATTTTTTTGCCATGGCGGGTGACTAAAAAATCTACAATGCGTATCTGCCCGCCTCTGTCGTGCAGCACGACGCCCGGGGGATAGCAGATCAGCTAGATTTCAAATATGCAAGAGAGGTTGTCCAGGCCAGACTGGGCCTGGGAAGCGTAAACTGGGTTTCCAGCTAACTCTAAGTGGAGCCATGATCATATCTCACTCCTGGAATAGCCCGAGATTTCTTGCCACATTAGTCTCATTTTTGAACAAGGAGTCCTAATTCTTAGGAGTTGAACTCACTGACCAATA\n",
      ">seq30\n",
      "TCGAAGTTCTCCACGGCGCATTTGACATAGTCGGTTCAATAATTTTGGTTACTTGTGGTACACCACACCTCCCGCCAAGGAATGGTTTTATCATTCTCAAAGAGGATTATCGGAAGATTTCGGAAAACATAGTTGTTAGAGTAAACAGTTTTGATCGCGTTGATAAGGTTTTCCACAAGAAGTGAAACGCTGAGTGACATAGGAATACCTTAACGCCCGATTTAGCGCTAGAGTACTTGCTGGTGGTCTAATATTCGTCGCAACGCGCGGTCTCTGCTAAAGTAGTGTCTGTGCCTGCCCCTGTCGCGCGGCGGGTCCTCGCCTGGTTACGTGGACCATCTGCCGCAGGGTATGCAAGATTAGTATGGATGCCGAAGCGCGAGCCCTTCCATAGTCGTAAAGA\n",
      ">seq31\n",
      "TCTCTCGCCTTCCGCATCTAGCTCAGATATGGTGTTATGAAGGCGTATCCTACTCGGGGAGCGACCGCCAATCCATGGGAGGATATCATACGGATCCCCCCTACCAGCCACATACCTTTGGCGGCTACGTGTAGATGAGACTTCAACATGCGTACACTAGATTTCAGGGATTATAAGGCGTTCTAAAAAAACGAAACGCTAAATTCCTAGTCAAAATCGAAAAAGCCTCATTCACAACAAGTCACGCGCTCTTAGGCCACTAGAATTCCCCCGCACTCAGCAAACTTGTGGCGTGAACTTTTCCTTGAAGTGGTCATTGTCACAATTGCGATCCGAACTTCATATACCATGTGAGTCTCGCCCATCTGAACGTTGAATTCTGGCTTCGCATGACTTGTCGGTCACTCACTTCTCACCGGGAGGTCGAGTCGGACGATAT\n",
      ">seq32\n",
      "ACCCCGCGTCTGCTGATTCATACGTACGACCCAACGGGGTTGCGGATTGTTATGACTCGTCCGGGCGGGCGATGACATCGGACCAGTCTGCCTGACTATTCTGCAATTCCGTGTTGCTTGCGCTACTCATGGTTGAGACGTGAAAAGAGAGACAAGTGCAAGAGTAGGTCGTACTTATCAGCGGAACGGCCGTGGGATATACAGGGACTAGGCGTTACTATAGATTGCTCGGCAACGCGCCTTGAATGCCGCGGATCATGGGTCCCCTTACTTTATAGACAATCAGTAACACCGGGTCGACCGTACACACGATTTTCAGCGCCCTATGAGGTGCGTCCATCAGACCTGGAGACCCGGCAC\n",
      ">seq33\n",
      "CACTGTCGAACTCTTCAGTGGTTATACGATTTCCGGTTCTTAGGATTAGTAGAGGTCATTTACTATGTCGCTAATTATACGGGCCCACGTATATCAGCGTGCGGTAATAAGTGAGGAGTCAATAGTGATACTTGTGAGCGCTTTAGGCAATTTCGCGTTCAGCCAGGAGGTTCGAACGCTATATCACGCCCCCGTTGTGTTTGTTAGTCCCTCTATTATTCCATGCGGCCCTGACCACAAAGTGATCATGGCGCCTTAGTGTCATACTAAAAGTCGCGCTCATGATCGGCCGGTACCTGAAGCCAGATCTCGGCGTGGGAGCTAATATCTGGCATCGTCAAATGGGTAAAGCTAACGAGCCATTTTATAGACAGTCCTTTGTTGGTGACCCGGGATCCTACTTACCTTGCACTCAGTC\n",
      ">seq34\n",
      "AACTGACGCGACGTTGGAAGCACACACCGGGCATCATCGGCGGCCACAACGTTAGTTCTCACCAACGGCGCGATCTATCTTTATGAGTACCAATCGAAAATCACTATGACTTTCAGCGGTGGATGCTGGCGAGATTGTGCAGCTCTGCATCTGCTTCGGACATGTTTTGACTAGGAAGGACGACTGATGAAAACTTCTCGAACGAAATGGACTTTCCTTGTGAGGAAACAGGCCGAGGAGCAAAACCGTAAACCCCGCGTCCGGGTCAAGAGACGGTATGCACCATCCGTGGTAATTGCGAGAACCTGCGACTCTCTGGTGTAAGTGGCGTGGTTCTTATCGCACCCTTGTTAGTATACTTGTTGGGATAGGGTTTACCTGGTATTGGTGGTGATCATGCTTAAATAGGCGGCCAGTCGATGGAAGTGCGGCCTCATGCGCGTTCTAGACGACCACGCCTTTATTGGGCTTACTAGACTAATTATGCATCGACCGTTTCTAAAAGGCTCTTGCAGCACACACGGAGCTCTA\n",
      ">seq35\n",
      "TCTAGTGAAACTCATCTCAGTACTCCGAGGGGGATAGGCGGACCGTTAGTAAGGACGCCCATTCATTGTGCTTTCACGCGAATATCTTACATTGTCAAGCCCCTCAACGCGCGGACGGGCGTCGGGGAGCAGCGTAAACGGACTTGGCTACGAACCCTGGGGGCAACGGGATGGGGCTCCGGCGGCACGTTCGTATGCTCTAGCGAGACGGGCCATACAAGTATGGAGCCGGTCTTTCCCGTCTGACCATGTACCGCCTTGCGCTCGCAGATCCCCAGAATTCCTTACAGCTTCGCTGGACCAGAGCGCCGTCTATGTAGGTCGCAGCCAAACCCGTCGATTAATTCTTTTTAGCTTGGGTGGCCACCATAGGAGTTTATCTTCCCTATAGAGTAAAGGACTCGAGGTTCTCACGCGTAAA\n",
      ">seq36\n",
      "CCAATGGGTGATGTGATTATTACCACCGTACCGCCGATCGCGGAGAGAAGTTGGATGGTGTAAACTTCATGAACCATTACTAGCATAGGAGTCAGACTGGCTTTACATCATAGGAGTGTCTTCTCGTAAAAGAAGGCTCACGGCAGGAGAAGAACCGAGCCCGTCGTGCTCCTGCATTGCAAAGCCAAAAATTCAGAGAGCTAGGAGTTTACTCTATGACTCTACAAGGAGTCCCTGGTTAATTTGACTGTGTCTGCATTCCAAGAAAGGGTCCTGGTACATGGAGACAGTAGCCCGCGCCCGTCTGACGACGCGCCAAGAGGGATGGGTGGGATCCAATACCTTAGAACATCCGGACGTCGGTCAACGCTTCTC\n",
      ">seq37\n",
      "GCTTAATGGGGACTTTGTATGGTTATCTACGTACTCGGATCCGTGTATAATTGTCGCCGATTCGGTGGACCTCTTTTCAGTGAGGAATTTGTGGTAATTTTGGACTCAAACTTTGCATCGGGCCTGGAACCTGACCGCTGTAACAGGACCTGCATCCATTCATACTCGGCTCGGGCGTGTGTGAGAGAAAGCATTATGGAACATCCTATCCATCCATCCGAACCGCAAACATATACCTAAGCGAGACTCCGGTTCGTCGTAACATGGTCATTCCAAAGCTCCGGTACCAGGAAGGTAGAGCAAGGTTCTGATCTACCGAATTAATCCATAACGCTAATTGGCCCGATTCCGGCTTAATTGATAGGGAATCCACTACGTTCGGCTCACGAACGACTACAGGACCCTGAGGCGGGGA\n",
      ">seq38\n",
      "TCAAAACGGTTGGAATTTATTTGAAAGGTCAGAATCTTAAAAGACGTGCCATCTCTCTCAATGACCGGACTTACCGAACAAAAGACACAGCGTTTCGGATTTAATTACGGAACCGCGAGCAGCTAAACCAAAGTGAGGGGCATCGACCTCCCTCGTGATGCGACTTTATCAACCGAGGGTGCTACGGTCGATCTACATACCGGCTTCCGTTGGGACACAGCTGCGCGCTGTGGATTTGGAATGTGCATGCCCACCCTTAGGCCATCACATGTTCCTAAACATTGCCGTCGCGGCTCTGCCGATAAGAAGCCCGTTGGGCCGGTCAAGACACATTCCTTTATTCCAGTGATCACGACTATGCCCTG\n",
      ">seq39\n",
      "GGTCTGCCGTTTAGACACTGTCATTAGGAGCGTGGGGCCCGCTCGATAGTGCTAGGTCCTGAGCGATAGTCGCGGAAACTTCAAAAACGCAATGGTTAAAACTTTCATGCACCTATATATTCCAAACATCGCCTGTTTTCGAAAGTGACCACTGCTTGCGCTACCGGAACTTCGAGTACCCATACAGCGGAAACGTTCAGGCAGCAGTTCTCGATAGGAATAAACCAAGACACAAGTCGACCCGCGACGGGCCACACGACTGTGTAAACCGGGATTGCGGGTAGGGCCTTTAGTCGGGATTAATGCGGTGCGAAATTTCATCCATCACGAATAAAACTGATTCCACTCCGAATCGGCTTCAACACTGGAGACTGCTTACGCGACTTCC\n",
      ">seq40\n",
      "CTTGGCACGGTCACTCATGCGTTTGTACCGTGGTTATGCATGGTCGTCCCGATTTTCGCAAATATACACCATAAAGGGTTCGTTCGACCTCCCCGAGGTTGCTAGGGTAGGGATTCACTAAGCCGTACCATGCCCGAGCTTCGGTGCAGCAGCAACCCGAGGCTTAGGTGAACCGTAACGGTCACACTTAGCTCCAAGTGAGCCGCACGAGCGCGAGCTTGTGGATCCATTTCTCGTACCTAAGAGCGGTGAATTCAAGCTGCTTTGGCCCCCTCAGCCCTCTTACGTGCGTCTTTGTGTGCGACATAGAAGCTAGTTCGAGGGAAAGTGGTGAACGTGATTCTAGTGCGGTTCGGGGTCTACGATGTGACCATGC\n",
      ">seq41\n",
      "GCCAAGCATGGTCTGAAGGACAGCCTAAACAAGAGCCCCCCTCGTGCCCGGGTCGTTCACTGTTATACCTGAAATAGCATAAACGTTATTTCAAGTCAAGTCGTAGATATGTGTTAATGCGTTAGGGAGGCTCGCAGTGTGCGATGATCCCTAGCATAATCTGCAATAAGACGCTTCCACGCATACACTGTATGACATTGATATTGACCCGTTGTACAAGTGGAGTTCCTACTGGGGCCTCGCGTCGTTCCTGTGTGGGCAAGCAACGCGCGTGGGTCATGTAGCGCATTAAACCGTTACAGTTTGAAAGCGGCTTAGGCAGGTAACCCCAGCTCTCATACGACCATGCCCATGCGAGGCAAAAACCAATTCGCAGTGCATCATCTCGTAGGTAACCCCGGGAAATAAATTTAACCGAACCTTGTAG\n",
      ">seq42\n",
      "AAACTTTGTTGTCCGACAAATTATGATCTGGTGCTCGAAGAATTCGACCAATGGGCTGAGAATCAACTCATAACTGTGCCAGTCGGTCATGATATGGAGGGTGCCGCGTGCCCGCCCACGTTAAAGCACCAGCCAACCTGTGGATGGTGCACAAACGATACCTCCGATCCTGCCCCCACCCGAAAGCTCGCACTGTTATGTCGGTTATCTCATTACAGTCGCTAGTTTAGGCTCCGCCATAGCTACTAGACGATATCTGCAAATAACTTACAGGCTGTGGTTCGGGTACAGGATCGTTACGGATTCGATTTATGTGACTGTGTAGAGTCAATCTCTGTTT\n",
      ">seq43\n",
      "GTCGAAGGCGATCCGCCGACGGGGTATCGTTGCCCGGATTCTTGGTCGACACTTCCCCTCAACTACCCGGAAACCGCTCAGCTACAACAGTCGAACCCATATGCAGGAGCCGGCGGTTCATTAACCTGCTAATCCGCGTGTCCACCCCTAATACTTATGTGTAGCAGGGTTACGGGCGCAGCCTCCCGGACAACTCTGAGTGTTCGTCCCGAACTTTTACTGTCGCAGTGATATAAAGCAAATAGTTTCAGAGCACGTATAGCGCCTCATTGTTCTTTACGCCTTTCACCATTGCAACTCCTGACAACGGGCGAGTTACCAAATAGTCAAAGGTTACTAGCGGTATAGTTGGCAAACCAT\n",
      ">seq44\n",
      "CAATACCCGCGTTAGAGATCCTGAACTAACTCTATAAGCGTCCACACTGGTACAAATCCGCTAAGCACAGTGTAGATCTAAGGCGCCCGGACGGGATTGGCATATGCTGAGGGTCTTAGGTCGTACAGGCGTCGAAGGTTGGGGGCGCCATCACGCAAGGACGGACGGTCAGACATATGGTCTTAATAAACCGCGGGAACGATCCAGGAGGCCGCGGAATGGGATTGCCTCTTATCTCCGCGGTCTCCAGTCGTTAGATTGAGTTGTATCGATCCTGTACACCCTCGCTGCTCATTACTTGAAACCGCTTCGTCGAAGGTATTGCAACGATACCACGATTCGTCGGTGGAACGACGCCAAACTCTGTTGCGTGCGGCAACAAGCCTACGTGGAGTGGGGGTCAAGATTGAAGCATATGTCAACACTGTCGATTAT\n",
      ">seq45\n",
      "TCGGCGCACGGTTATCATCGGACGCCGTATGTTGACCGGGGCGCCGCCAACAATTGCCGTACGCTTCTACTAGGAGTGCTGATGTACTATGTCCTACAGACGAGTGTCTGGCGCCGAAGGACTCAAAAGCGAATGCCGAATGGTTGAATACCACTAACTCTACACTATACAGTACATCAGTAGCGCAAACGCTTTATTATCTTAGTTCCGCTGAGAAGTTCTTCCAAGTGCAGGCGATAGGTTGTTTGCGTCATATTGCGCAACTGGATGGTTCGTAAGTCAAGCACTATGACGGCGCACTGTTCCCGGCAAGAAAGTGACGACGCGGTAGCTAATGAATTGTCTTAGCATAACACCAGAAATCGATACCGGATGAACGCGGACTGAGATCCGGTCGGCTCTCCGAAGTACTGCTGACTTACGTTACTTACTACGTTCGTGGTTTGCTA\n",
      ">seq46\n",
      "CGCTTCACTTGTTACAAACTGCTGTGTTGGAAGCAGTACTTCTAACAGAAGGTTGCACGTGGTTGAAGTTAGTCGCACTCACTGACCGTTAATGTGGCTCCCGAACGGTAGTTATTCATCAGTTGAGGGCTCTTAACATTTAGTCCCCACGACACGAGTGTGCATGAGTCCTTGGCTCAGTTCTAGGATATCTAATTCGTCGTTATATCACCAGGCTATTGGTTCAGCAAATGGCACACCGCGGTGGTGTCGTGGTCAAGTTAACGGGAACGTCCTCTCAGTAAGGGTTTATGCTGTGTCGTTAGATCGTACAGACTGAAGCCCGTGATTGCAGGCGACTAATCTCCTGTCACCTAGGATCTAGCAGTGGCGCATACCATAGTGACACAGAGCCTGAGAAGCCGCCGCAATTTTTTTCATTAGCC\n",
      ">seq47\n",
      "GGATAGTTATTCCGAAATAGGTTCACTCACCTCACATTGCTATAAGCCTACATCATTTAGCTAGTGATCAGACGGAAAAGATAAATCACATTTGGAGAAGGGCTTACGACAGCCTTACCATAACAATACTGTAAAGCCACACGTACCGCGCCCGCTCGGCGTATTGGTCCACGGGTCTCGGACCACTATCTCGAACTGTCCAGAGGGGGGCAACAAAGGCAATGTAGCATCGACCATATTTGAACCCTAGGATCGACATTTCGTAAACTTTTCGAGCAAGACAGAAGACACGCACACTTCCGTACGATGTTGACGGCCTTACCCTGGCGCCATAGCATCTAGTGGTAGGAGATGTGTGCTTACGGCTCAGAAGTTCGCGATATGCGTGCAGAAGAAATACCGCAGTGGAAGGAGTACTCTAGATAATTGTAACCAATCTCCGGCGCAGCAATTTCGGA\n",
      ">seq48\n",
      "CTATGCACAACGTAGCTTCGGGCAAGGCTTTCTGAGAACCTGGCAGCATCAGCTCCAATCGATTTCATGACACTAGGACGGCGCTTAGTTCGCCTCCCTCATCACACCAAGTTTCCTCTGTCCGTTCTAGGAACACGCCCTTTGCTGGCCCCCCCCGTCGTTAGTCGGCATACAGTCTGTACTGAACTAGGCGTGCTCGAGACTTGTCTACAGGTTCAGCAACATTGGTGGATTTAAGAGGCCGATCGATCAGTGCTCCCCTCTCCTACTAGCAGGGAAGACAAGGGAAGCCGGCTGTTGAATATATTCAATTATAGATTATTCTCGGGACTCAAACCTACGTCAGTTAGCAGACACACGCAAAGTTCT\n",
      ">seq49\n",
      "TACGAAATCTTATTTTACACATCCACACGCCTACTTACAGGTATGCAACTTTGGGAATCCACCACCAGGTCCCAGCTCCACGTTAGGTTAACTCCGTGCGCCGCGGCCCACTGAGAATCATTCCGCGGTCGAACTTAGACATGACTCCCGCAGGTCTGCGGTCGTTGGCTCAGTGTGAGTTCGAAATGAGGGGGCCTGTGCCAGAAGAGCTTCTCTCTGAGCTGCTCTGTCAGTCAACCGACAAAGGACTGACGGTTACCCAGTGGGCATGTTGTCCTTTCCGCACCAGGGTGGATATACGGGCACCGGGACTTCATAGCTGAACGTCTTGACTATGCGGGTAGCGACCCTCGAGAAGGTGCTATCGCTACCTAAAACGTTAGCAGAGGATGCTGTATAT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(trimmed_sequences)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
